File: 2023/01/lua/main.lua
#!/usr/local/bin/lua

local run = require('src.lua.run')
local str = require('src.lua.utils.str')

local options = {
   has_alternate = true,
   has_io = false,
}

local function is_num(c)
   return c >= '0' and c <= '9'
end

local function get_num(s)
   local c = s:sub(1, 1)
   if is_num(c) then
      return tonumber(c)
   end
   c = s
   if str.starts_with(c, 'zero') then
      return '0'
   end
   if str.starts_with(c, 'one') then
      return '1'
   end
   if str.starts_with(c, 'two') then
      return '2'
   end
   if str.starts_with(c, 'three') then
      return '3'
   end
   if str.starts_with(c, 'four') then
      return '4'
   end
   if str.starts_with(c, 'five') then
      return '5'
   end
   if str.starts_with(c, 'six') then
      return '6'
   end
   if str.starts_with(c, 'seven') then
      return '7'
   end
   if str.starts_with(c, 'eight') then
      return '8'
   end
   if str.starts_with(c, 'nine') then
      return '9'
   end
   return nil
end

---@param input string
---@param is_test boolean
---@return string
local function part_1(input, is_test)
   local lines = {}
   for line in input:gmatch('[^\n]+') do
      table.insert(lines, line)
   end
   local res = 0

   for i = 1, #lines do
      local line = lines[i]
      local first, last

      for j = 1, #line do
         local c = line:sub(j, j)
         if is_num(c) then
            first = c
            break
         end
      end
      for j = #line, 1, -1 do
         local c = line:sub(j, j)
         if is_num(c) then
            last = c
            break
         end
      end

      res = res + tonumber(first .. last)
   end

   return tostring(res)
end

---@param input string
---@param is_test boolean
---@return string
local function part_2(input, is_test)
   local lines = {}
   for line in input:gmatch('[^\n]+') do
      table.insert(lines, line)
   end
   local res = 0

   for i = 1, #lines do
      local line = lines[i]
      local first, last

      for j = 1, #line do
         local c = get_num(line:sub(j, #line))
         if c ~= nil then
            first = c
            break
         end
      end
      for j = #line, 1, -1 do
         local c = get_num(line:sub(j, #line))
         if c ~= nil then
            last = c
            break
         end
      end

      res = res + tonumber(first .. last)
   end

   return tostring(res)
end

local function main()
   run(arg, part_1, part_2, options)
end

main()

File: 2023/01/ocaml/main.ml
let options = Hashtbl.create 128;;

Hashtbl.add options "has_alternate" true;;
Hashtbl.add options "has_io" false

let is_digit = function
  | '0' .. '9' -> true
  | _ -> false
;;

let fetch_num_p1 (s : string) : int =
  let nums = s |> String.to_seq |> Seq.filter is_digit |> Array.of_seq in
  int_of_string (String.make 1 nums.(0) ^ String.make 1 nums.(Array.length nums - 1))
;;

let part_1 (input : string) (_is_test : bool) : string =
  input
  |> String.split_on_char '\n'
  |> List.map fetch_num_p1
  |> List.fold_left ( + ) 0
  |> string_of_int
;;

let get_num (s : string) : char =
  if String.starts_with ~prefix:"zero" s
  then '0'
  else if String.starts_with ~prefix:"one" s
  then '1'
  else if String.starts_with ~prefix:"two" s
  then '2'
  else if String.starts_with ~prefix:"three" s
  then '3'
  else if String.starts_with ~prefix:"four" s
  then '4'
  else if String.starts_with ~prefix:"five" s
  then '5'
  else if String.starts_with ~prefix:"six" s
  then '6'
  else if String.starts_with ~prefix:"seven" s
  then '7'
  else if String.starts_with ~prefix:"eight" s
  then '8'
  else if String.starts_with ~prefix:"nine" s
  then '9'
  else ' '
;;

let rec fetch_num s i r : char =
  if s <> ""
  then (
    let c = String.get s i in
    let d = String.sub s i (String.length s - i) |> get_num in
    if is_digit c
    then c
    else if is_digit d
    then d
    else fetch_num s (if r then i - 1 else i + 1) r)
  else ' '
;;

let fetch_num_p2 (s : string) : int =
  let first : char = fetch_num s 0 false in
  let last : char = fetch_num s (String.length s - 1) true in
  int_of_string (String.make 1 first ^ String.make 1 last)
;;

let part_2 (input : string) (_is_test : bool) : string =
  input
  |> String.split_on_char '\n'
  |> List.map fetch_num_p2
  |> List.fold_left ( + ) 0
  |> string_of_int
;;

let () = Aoc.run Sys.argv part_1 part_2 options
